File: frontend/src/utils/setupConfig.js
import { DAYS_OF_WEEK, DEFAULT_WORKING_DAYS } from './constants';

/**
 * setupConfig.js - Pure helpers for the Setup (timetable configuration) feature.
 *
 * All time values are stored internally as 24-hour "HH:MM" strings so that
 * arithmetic is unambiguous. 12-hour AM/PM rendering happens only at display
 * time via `formatTime` / `formatRange`.
 *
 * This module is deliberately UI-agnostic so a Supabase-backed persistence
 * layer can later replace `loadSetupConfig` / `saveSetupConfig` without
 * touching the components.
 *
 * NEW DATA MODEL:
 * - breakAfter: number[] - teaching period numbers AFTER which break occurs
 * - lunchAfter: number | null - teaching period number AFTER which lunch occurs
 * - numberOfPeriods: number - TEACHING periods only (breaks/lunch don't count)
 *
 * Old model (for migration):
 * - breakPeriods: number[] - period numbers treated as breaks
 * - lunchBreakPeriod: number | null - period number treated as lunch
 */

export const MAX_PERIODS = 15;
export const MIN_PERIODS = 1;
export const MAX_DURATION_MINUTES = 180;
export const MAX_BREAK_MINUTES = 120;
export const MAX_LUNCH_MINUTES = 120;

const STORAGE_KEY = 'setup_config';
const TIME_REGEX = /^\d{2}:\d{2}$/;

/* ----------------------------------------------------------------
   Numeric / time utilities
---------------------------------------------------------------- */

export function clampInt(value, min, max) {
  const n = Number.parseInt(value, 10);
  if (Number.isNaN(n)) return min;
  return Math.min(max, Math.max(min, n));
}

/** Convert "HH:MM" to minutes since midnight. Returns NaN when invalid. */
export function timeToMinutes(time) {
  if (!time || typeof time !== 'string') return Number.NaN;
  const match = time.match(/^(\d{1,2}):(\d{2})$/);
  if (!match) return Number.NaN;
  const hours = Number(match[1]);
  const minutes = Number(match[2]);
  if (hours > 23 || minutes > 59) return Number.NaN;
  return hours * 60 + minutes;
}

/** Convert minutes since midnight back to "HH:MM", wrapping across midnight. */
export function minutesToTime(minutes) {
  const total = ((Math.trunc(minutes) % 1440) + 1440) % 1440;
  const h = Math.floor(total / 60);
  const m = total % 60;
  return `${String(h).padStart(2, '0')}:${String(m).padStart(2, '0')}`;
}

/** Add minutes to a "HH:MM" string, wrapping past midnight. */
export function addMinutes(time, minutes) {
  const base = timeToMinutes(time);
  if (Number.isNaN(base)) return time;
  return minutesToTime(base + minutes);
}

/** End - start in minutes. Returns 0 when either value is invalid. */
export function diffMinutes(startTime, endTime) {
  const start = timeToMinutes(startTime);
  const end = timeToMinutes(endTime);
  if (Number.isNaN(start) || Number.isNaN(end)) return 0;
  return end - start;
}

export function isValidTime(time) {
  return TIME_REGEX.test(time) && !Number.isNaN(timeToMinutes(time));
}

/** "09:00" -> "09:00 AM", "16:00" -> "04:00 PM" */
export function formatTime(time) {
  const minutes = timeToMinutes(time);
  if (Number.isNaN(minutes)) return time || '--:--';
  const hours = Math.floor(minutes / 60) % 12 || 12;
  const mins = String(minutes % 60).padStart(2, '0');
  const meridian = minutes < 720 ? 'AM' : 'PM';
  return `${String(hours).padStart(2, '0')}:${mins} ${meridian}`;
}

/** "09:00 AM - 09:50 AM" */
export function formatRange(startTime, endTime) {
  return `${formatTime(startTime)} - ${formatTime(endTime)}`;
}

/** Compact "09:00 AM–09:50 AM" used inside table headers. */
export function formatShortRange(startTime, endTime) {
  return `${formatTime(startTime)}\u2013${formatTime(endTime)}`;
}

/** "9th", "1st", "2nd", "3rd" */
export function ordinal(n) {
  const suffixes = ['th', 'st', 'nd', 'rd'];
  const v = n % 100;
  return n + (suffixes[(v - 20) % 10] || suffixes[v] || suffixes[0]);
}

/** "9th Period" */
export function periodLabel(n) {
  return `${ordinal(n)} Period`;
}

/* ----------------------------------------------------------------
   Period derivation
----------------------------------------------------------------- */

/**
 * Build a contiguous run of periods starting at `startTime`.
 * 
 * NEW MODEL:
 * - `count` = number of TEACHING periods only
 * - `breakAfter` = array of teaching period numbers AFTER which break occurs
 * - `lunchAfter` = teaching period number AFTER which lunch occurs
 * - Break/lunch are intervals BETWEEN teaching periods, not periods themselves
 * 
 * Example: 8 teaching periods, breakAfter=[2], lunchAfter=[5]
 * Timeline: 1, 2, BREAK, 3, 4, 5, LUNCH, 6, 7, 8
 * Total displayed slots = 8 teaching + 1 break + 1 lunch = 10 slots
 * But teaching period count = 8
 */
export function derivePeriods(startTime, duration, count, options = {}) {
  const {
    breakAfter = [],
    lunchAfter = null,
    breakTime = 0,
    lunchTime = 0,
  } = options;
  
  const breakAfterSet = new Set(breakAfter);
  const periods = [];
  let start = startTime;
  
  for (let i = 1; i <= count; i += 1) {
    // Add teaching period
    const end = addMinutes(start, duration);
    periods.push({
      periodNumber: i,
      startTime: start,
      endTime: end,
      totalMinutes: diffMinutes(start, end),
      isTeaching: true,
    });
    start = end;
    
    // Add break interval AFTER this teaching period if configured
    if (breakAfterSet.has(i)) {
      const breakEnd = addMinutes(start, breakTime);
      periods.push({
        periodNumber: i, // Same number as the period before break
        startTime: start,
        endTime: breakEnd,
        totalMinutes: diffMinutes(start, breakEnd),
        isTeaching: false,
        isBreak: true,
      });
      start = breakEnd;
    }
    
    // Add lunch interval AFTER this teaching period if configured
    if (lunchAfter === i) {
      const lunchEnd = addMinutes(start, lunchTime);
      periods.push({
        periodNumber: i, // Same number as the period before lunch
        startTime: start,
        endTime: lunchEnd,
        totalMinutes: diffMinutes(start, lunchEnd),
        isTeaching: false,
        isLunch: true,
      });
      start = lunchEnd;
    }
  }
  return periods;
}

/**
 * Get only the teaching periods from a derived periods array.
 * Useful for counting actual teaching periods.
 */
export function getTeachingPeriods(periods) {
  return periods.filter(p => p.isTeaching);
}

/**
 * Get the count of teaching periods.
 */
export function getTeachingPeriodCount(periods) {
  return periods.filter(p => p.isTeaching).length;
}

/** Rebuild a single period entry from its start/end times. */
export function makePeriod(periodNumber, startTime, endTime, isTeaching = true, isBreak = false, isLunch = false) {
  const validStart = isValidTime(startTime) ? startTime : '';
  const validEnd = isValidTime(endTime) ? endTime : '';
  return {
    periodNumber,
    startTime: validStart,
    endTime: validEnd,
    totalMinutes: diffMinutes(validStart, validEnd),
    isTeaching,
    isBreak,
    isLunch,
  };
}

/* ----------------------------------------------------------------
   Defaults
----------------------------------------------------------------- */

export function buildDefaultConfig() {
  const numberOfPeriods = 9;
  const collegeStartTime = '09:00';
  const collegeEndTime = '16:00';
  const periodDuration = 50;
  const breakTime = 10;
  const lunchTime = 40;
  const breakAfter = [3]; // After 3rd period
  const lunchAfter = 6;   // After 6th period

  return {
    workingDays: [...DEFAULT_WORKING_DAYS],
    numberOfPeriods,
    collegeStartTime,
    collegeEndTime,
    periodDuration,
    breakTime,
    lunchTime,
    breakAfter,
    lunchAfter,
    periods: derivePeriods(collegeStartTime, periodDuration, numberOfPeriods, {
      breakAfter,
      lunchAfter,
      breakTime,
      lunchTime,
    }),
  };
}

/**
 * Convert a persisted config document (which may use snake_case keys, as
 * produced by external clients) into the camelCase shape `sanitizeConfig`
 * understands. Unknown/missing keys are left as undefined so sanitizeConfig
 * falls back to defaults.
 */
export function configToCamelCase(raw) {
  if (!raw || typeof raw !== 'object') return null;

  const periods = Array.isArray(raw.periods)
    ? raw.periods.map((p) => {
        if (!p || typeof p !== 'object') return null;
        return {
          periodNumber: p.periodNumber ?? p.period_number,
          startTime: p.startTime ?? p.start_time,
          endTime: p.endTime ?? p.end_time,
          totalMinutes: p.totalMinutes ?? p.total_minutes,
          isTeaching: p.isTeaching ?? true,
          isBreak: p.isBreak ?? false,
          isLunch: p.isLunch ?? false,
        };
      })
    : undefined;

  // Handle both old (breakPeriods, lunchBreakPeriod) and new (breakAfter, lunchAfter) keys
  const breakAfter = raw.breakAfter ?? raw.break_after ?? 
    (Array.isArray(raw.breakPeriods) ? raw.breakPeriods.map(p => p) : []);
  const lunchAfter = raw.lunchAfter ?? raw.lunch_after ?? raw.lunchBreakPeriod ?? raw.lunch_break_period;

  return {
    workingDays: raw.workingDays ?? raw.working_days,
    numberOfPeriods: raw.numberOfPeriods ?? raw.number_of_periods,
    collegeStartTime: raw.collegeStartTime ?? raw.college_start_time,
    collegeEndTime: raw.collegeEndTime ?? raw.college_end_time,
    periodDuration: raw.periodDuration ?? raw.period_duration,
    breakTime: raw.breakTime ?? raw.break_time,
    lunchTime: raw.lunchTime ?? raw.lunch_time,
    breakAfter,
    lunchAfter,
    periods,
  };
}

/**
 * Merge an untrusted persisted object with the defaults, clamping every value
 * and re-deriving anything that is stale or malformed.
 *
 * MIGRATION: Supports both old model (breakPeriods, lunchBreakPeriod) and
 * new model (breakAfter, lunchAfter).
 */
export function sanitizeConfig(raw) {
  if (!raw || typeof raw !== 'object') return buildDefaultConfig();

  const numberOfPeriods = clampInt(raw.numberOfPeriods, MIN_PERIODS, MAX_PERIODS);
  const workingDays = (Array.isArray(raw.workingDays) ? raw.workingDays : []).filter((day) =>
    DAYS_OF_WEEK.includes(day)
  );
  const days = workingDays.length ? workingDays : [...DEFAULT_WORKING_DAYS];

  const periodDuration = clampInt(raw.periodDuration, 1, MAX_DURATION_MINUTES);
  const breakTime = clampInt(raw.breakTime, 1, MAX_BREAK_MINUTES);
  const lunchTime = clampInt(raw.lunchTime, 1, MAX_LUNCH_MINUTES);
  const collegeStartTime = isValidTime(raw.collegeStartTime) ? raw.collegeStartTime : '09:00';
  const collegeEndTime = isValidTime(raw.collegeEndTime) ? raw.collegeEndTime : '16:00';

  // Migration: old model used breakPeriods (periods treated as break),
  // new model uses breakAfter (periods AFTER which break occurs).
  // For migration, convert old breakPeriods to breakAfter.
  let breakAfter = [];
  if (Array.isArray(raw.breakAfter)) {
    breakAfter = raw.breakAfter
      .map(Number)
      .filter((p) => Number.isInteger(p) && p >= MIN_PERIODS && p < numberOfPeriods)
      .sort((a, b) => a - b);
  } else if (Array.isArray(raw.breakPeriods)) {
    // Old model: breakPeriods meant "this period IS a break".
    // New model: breakAfter means "break AFTER this period".
    // For migration, convert: if break was period 3, now break after period 2.
    breakAfter = raw.breakPeriods
      .map(Number)
      .map((p) => p - 1) // Convert: period that WAS break -> period BEFORE which break was
      .filter((p) => Number.isInteger(p) && p >= MIN_PERIODS && p < numberOfPeriods)
      .sort((a, b) => a - b);
  }

  // Migration: old model used lunchBreakPeriod, new model uses lunchAfter.
  let lunchAfter = null;
  if (raw.lunchAfter != null) {
    const lunchCandidate = Number(raw.lunchAfter);
    if (Number.isInteger(lunchCandidate) && lunchCandidate >= MIN_PERIODS && lunchCandidate < numberOfPeriods) {
      lunchAfter = lunchCandidate;
    }
  } else if (raw.lunchBreakPeriod != null) {
    // Old model: lunchBreakPeriod meant "this period IS lunch".
    // New model: lunchAfter means "lunch AFTER this period".
    const lunchCandidate = Number(raw.lunchBreakPeriod) - 1;
    if (Number.isInteger(lunchCandidate) && lunchCandidate >= MIN_PERIODS && lunchCandidate < numberOfPeriods) {
      lunchAfter = lunchCandidate;
    }
  }

  // The periods array may contain old-format entries (without isTeaching flags)
  // or new-format entries. If stored periods don't have isTeaching, we should
  // still accept them but re-derive to get correct timings.
  const storedPeriods =
    Array.isArray(raw.periods) && raw.periods.length > 0
      ? raw.periods.map((p, index) =>
          makePeriod(
            p?.periodNumber ?? p?.period_number ?? (index + 1),
            p?.startTime || p?.start_time || '',
            p?.endTime || p?.end_time || '',
            p?.isTeaching ?? true,
            p?.isBreak ?? false,
            p?.isLunch ?? false,
          )
        )
      : null;

  const periods =
    storedPeriods && storedPeriods.some((p) => isValidTime(p.startTime) && isValidTime(p.endTime))
      ? storedPeriods
      : derivePeriods(collegeStartTime, periodDuration, numberOfPeriods, {
          breakAfter,
          lunchAfter,
          breakTime,
          lunchTime,
        });

  return {
    workingDays: days,
    numberOfPeriods,
    collegeStartTime,
    collegeEndTime,
    periodDuration,
    breakTime,
    lunchTime,
    breakAfter,
    lunchAfter,
    periods,
  };
}

/* ----------------------------------------------------------------
   Persistence seam (swap for Supabase later)
---------------------------------------------------------------- */

export function loadSetupConfig() {
  try {
    const raw = JSON.parse(localStorage.getItem(STORAGE_KEY));
    return sanitizeConfig(raw);
  } catch {
    return buildDefaultConfig();
  }
}

export function saveSetupConfig(config) {
  try {
    localStorage.setItem(STORAGE_KEY, JSON.stringify(config));
    return true;
  } catch {
    return false;
  }
}

export function clearSetupConfig() {
  try {
    localStorage.removeItem(STORAGE_KEY);
  } catch {
    /* storage unavailable - ignore */
  }
}

/* ----------------------------------------------------------------
   Validation
---------------------------------------------------------------- */

/**
 * Validate the whole configuration and produce field-level errors plus a list
 * of non-blocking warnings. Returns:
 *   { valid, errors: {field: message}, warnings: [string] }
 */
export function validateSetupConfig(config) {
  const errors = {};
  const warnings = [];

  if (!config.workingDays || config.workingDays.length === 0) {
    errors.workingDays = 'At least one working day must be selected.';
  }

  if (!config.numberOfPeriods || config.numberOfPeriods < MIN_PERIODS) {
    errors.numberOfPeriods = 'At least one period is required.';
  } else if (config.numberOfPeriods > MAX_PERIODS) {
    errors.numberOfPeriods = `A maximum of ${MAX_PERIODS} periods is allowed.`;
  }

  if (!isValidTime(config.collegeStartTime)) {
    errors.collegeStartTime = 'College start time is required.';
  }
  if (!isValidTime(config.collegeEndTime)) {
    errors.collegeEndTime = 'College end time is required.';
  }
  if (
    isValidTime(config.collegeStartTime) &&
    isValidTime(config.collegeEndTime) &&
    timeToMinutes(config.collegeEndTime) <= timeToMinutes(config.collegeStartTime)
  ) {
    errors.collegeEndTime = 'College end time must be later than start time.';
  }

  if (!config.periodDuration || config.periodDuration <= 0) {
    errors.periodDuration = 'Period duration must be greater than 0 minutes.';
  }

  if (!config.breakTime || config.breakTime <= 0) {
    errors.breakTime = 'Break time must be greater than 0 minutes.';
  }
  if (!config.lunchTime || config.lunchTime <= 0) {
    errors.lunchTime = 'Lunch time must be greater than 0 minutes.';
  }

  // Validate breakAfter and lunchAfter using the new model
  const breakAfter = Array.isArray(config.breakAfter) ? config.breakAfter : [];
  const lunchAfter = config.lunchAfter;

  // Rule: Break and Lunch cannot be after the same period
  if (lunchAfter != null && breakAfter.includes(lunchAfter)) {
    errors.breakAfter = 'Break and Lunch cannot both be scheduled after the same period.';
  }

  // Rule: Break/Lunch after the final period is not allowed
  if (lunchAfter != null && lunchAfter >= config.numberOfPeriods) {
    errors.lunchAfter = 'Lunch must be scheduled before the final period.';
  }
  for (const b of breakAfter) {
    if (b >= config.numberOfPeriods) {
      errors.breakAfter = 'Break must be scheduled before the final period.';
      break;
    }
  }

  // Validate period timings (teaching periods only)
  const periods = Array.isArray(config.periods) ? config.periods : [];
  const teachingPeriods = periods.filter(
    (p) => p.isTeaching === true || p.isTeaching === undefined
  );

  for (const period of teachingPeriods) {
    if (!isValidTime(period.startTime) || !isValidTime(period.endTime)) {
      errors.periods = `Please set valid start and end times for ${periodLabel(period.periodNumber)}.`;
      break;
    }
    if (timeToMinutes(period.endTime) <= timeToMinutes(period.startTime)) {
      errors.periods = `${periodLabel(period.periodNumber)} end time must be later than its start time.`;
      break;
    }
  }

  // Check teaching periods don't overlap
  if (!errors.periods) {
    for (let i = 0; i < teachingPeriods.length; i += 1) {
      for (let j = i + 1; j < teachingPeriods.length; j += 1) {
        const a = teachingPeriods[i];
        const b = teachingPeriods[j];
        const aStart = timeToMinutes(a.startTime);
        const aEnd = timeToMinutes(a.endTime);
        const bStart = timeToMinutes(b.startTime);
        const bEnd = timeToMinutes(b.endTime);
        if (aStart < bEnd && bStart < aEnd) {
          errors.periods = `${periodLabel(a.periodNumber)} and ${periodLabel(b.periodNumber)} timings overlap.`;
          break;
        }
      }
      if (errors.periods) break;
    }
  }

  // Warn if last teaching period extends beyond college end time
  if (!errors.periods && teachingPeriods.length > 0) {
    const lastPeriod = teachingPeriods[teachingPeriods.length - 1];
    if (
      lastPeriod &&
      isValidTime(lastPeriod.endTime) &&
      isValidTime(config.collegeEndTime) &&
      timeToMinutes(lastPeriod.endTime) > timeToMinutes(config.collegeEndTime)
    ) {
      warnings.push('The last period extends beyond the college end time.');
    }
  }

  return { valid: Object.keys(errors).length === 0, errors, warnings };
}

File: backend/app/services/setup_config.py
"""
Purpose: Setup Configuration Service
Author: Smart Timetable Backend Team
Module Description: Normalizes the raw Setup configuration document (stored as
JSONB and produced by the frontend with camelCase keys) into a canonical
snake_case structure the scheduler, exports and APIs can consume. Missing or
malformed values fall back to sensible defaults so the app keeps working even
before the user saves a Setup.

NEW DATA MODEL:
- break_after: list[int] - teaching period numbers AFTER which break occurs
- lunch_after: int | None - teaching period number AFTER which lunch occurs
- number_of_periods: int - TEACHING periods only (breaks/lunch don't count)

Old model (for migration):
- break_periods: list[int] - period numbers treated as breaks
- lunch_break_period: int | None - period number treated as lunch
"""

from typing import Any, Dict, List, Optional

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.setup_config_model import SetupConfig
from app.utils.exceptions import SchedulingFailureError

DEFAULT_WORKING_DAYS = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday"]
DEFAULT_COLLEGE_START = "09:00"
DEFAULT_COLLEGE_END = "16:00"
DEFAULT_PERIOD_DURATION = 50
DEFAULT_BREAK_TIME = 10
DEFAULT_LUNCH_TIME = 40
DEFAULT_NUMBER_OF_PERIODS = 9


def _time_to_minutes(value: str) -> Optional[int]:
    """Convert 'HH:MM' to minutes since midnight, or None when invalid."""
    if not value or not isinstance(value, str):
        return None
    parts = value.split(":")
    if len(parts) != 2:
        return None
    try:
        hours = int(parts[0])
        minutes = int(parts[1])
    except (TypeError, ValueError):
        return None
    if hours > 23 or minutes > 59:
        return None
    return hours * 60 + minutes


def _minutes_to_time(minutes: int) -> str:
    minutes = ((int(minutes) % 1440) + 1440) % 1440
    hours = minutes // 60
    mins = minutes % 60
    return f"{hours:02d}:{mins:02d}"


def _add_minutes(time_str: str, minutes: int) -> str:
    base = _time_to_minutes(time_str)
    if base is None:
        return time_str
    return _minutes_to_time(base + minutes)


def _is_valid_time(value: Any) -> bool:
    return isinstance(value, str) and _time_to_minutes(value) is not None


def _get(obj: Dict[str, Any], *keys: str, default: Any = None) -> Any:
    """Read a key tolerating both snake_case and camelCase spellings."""
    for key in keys:
        if key in obj and obj[key] is not None:
            return obj[key]
    return default


def build_default_config() -> Dict[str, Any]:
    """
    Return the canonical default configuration, mirroring the frontend
    buildDefaultConfig defaults.
    """
    periods = derive_periods(
        DEFAULT_COLLEGE_START,
        DEFAULT_PERIOD_DURATION,
        DEFAULT_NUMBER_OF_PERIODS,
        break_after=[3],
        lunch_after=6,
        break_time=DEFAULT_BREAK_TIME,
        lunch_time=DEFAULT_LUNCH_TIME,
    )
    return {
        "working_days": list(DEFAULT_WORKING_DAYS),
        "number_of_periods": DEFAULT_NUMBER_OF_PERIODS,
        "college_start_time": DEFAULT_COLLEGE_START,
        "college_end_time": DEFAULT_COLLEGE_END,
        "period_duration": DEFAULT_PERIOD_DURATION,
        "break_time": DEFAULT_BREAK_TIME,
        "lunch_time": DEFAULT_LUNCH_TIME,
        "break_after": [3],
        "lunch_after": 6,
        "periods": periods,
    }


def derive_periods(
    start_time: str,
    duration: int,
    count: int,
    break_after: List[int] = None,
    lunch_after: Optional[int] = None,
    break_time: int = 0,
    lunch_time: int = 0,
) -> List[Dict[str, Any]]:
    """
    Build a contiguous timeline of periods starting at `start_time`.

    NEW MODEL: count = number of TEACHING periods only.
    break_after = list of teaching period numbers AFTER which break occurs.
    lunch_after = teaching period number AFTER which lunch occurs.
    Break/lunch are intervals BETWEEN teaching periods, not periods themselves.

    Example: 8 teaching periods, break_after=[2], lunch_after=5
    Timeline: 1, 2, BREAK, 3, 4, 5, LUNCH, 6, 7, 8
    """
    break_after = break_after or []
    break_set = set(break_after)
    periods = []
    start = start_time

    for i in range(1, count + 1):
        # Teaching period
        end = _add_minutes(start, duration)
        periods.append(
            {
                "period_number": i,
                "start_time": start,
                "end_time": end,
                "total_minutes": max(_time_to_minutes(end) - _time_to_minutes(start), 0),
                "is_teaching": True,
                "is_break": False,
                "is_lunch": False,
            }
        )
        start = end

        # Break interval AFTER this teaching period
        if i in break_set:
            break_end = _add_minutes(start, break_time)
            periods.append(
                {
                    "period_number": i,
                    "start_time": start,
                    "end_time": break_end,
                    "total_minutes": max(_time_to_minutes(break_end) - _time_to_minutes(start), 0),
                    "is_teaching": False,
                    "is_break": True,
                    "is_lunch": False,
                }
            )
            start = break_end

        # Lunch interval AFTER this teaching period
        if lunch_after is not None and i == lunch_after:
            lunch_end = _add_minutes(start, lunch_time)
            periods.append(
                {
                    "period_number": i,
                    "start_time": start,
                    "end_time": lunch_end,
                    "total_minutes": max(_time_to_minutes(lunch_end) - _time_to_minutes(start), 0),
                    "is_teaching": False,
                    "is_break": False,
                    "is_lunch": True,
                }
            )
            start = lunch_end

    return periods


def normalize_config(raw: Optional[Dict[str, Any]]) -> Dict[str, Any]:
    """
    Sanitize an untrusted config document (frontend camelCase or backend
    snake_case) into the canonical snake_case structure used everywhere in the
    backend. Never raises for malformed input; defaults are applied instead.

    Supports migration from old model (break_periods, lunch_break_period)
    to new model (break_after, lunch_after).
    """
    if not raw or not isinstance(raw, dict):
        return build_default_config()

    number_of_periods = _get(raw, "number_of_periods", "numberOfPeriods", default=DEFAULT_NUMBER_OF_PERIODS)
    try:
        number_of_periods = int(number_of_periods)
    except (TypeError, ValueError):
        number_of_periods = DEFAULT_NUMBER_OF_PERIODS
    number_of_periods = max(1, min(number_of_periods, 15))

    working_days = _get(raw, "working_days", "workingDays", default=[])
    if not isinstance(working_days, list) or not working_days:
        working_days = list(DEFAULT_WORKING_DAYS)
    else:
        working_days = [d for d in working_days if d in DEFAULT_WORKING_DAYS]
        if not working_days:
            working_days = list(DEFAULT_WORKING_DAYS)

    def _clamp(value, default, minimum, maximum):
        try:
            return max(minimum, min(int(value), maximum))
        except (TypeError, ValueError):
            return default

    period_duration = _clamp(
        _get(raw, "period_duration", "periodDuration"), DEFAULT_PERIOD_DURATION, 1, 180
    )
    break_time = _clamp(_get(raw, "break_time", "breakTime"), DEFAULT_BREAK_TIME, 1, 120)
    lunch_time = _clamp(_get(raw, "lunch_time", "lunchTime"), DEFAULT_LUNCH_TIME, 1, 120)
    college_start_time = _get(raw, "college_start_time", "collegeStartTime", default=DEFAULT_COLLEGE_START)
    if not _is_valid_time(college_start_time):
        college_start_time = DEFAULT_COLLEGE_START
    college_end_time = _get(raw, "college_end_time", "collegeEndTime", default=DEFAULT_COLLEGE_END)
    if not _is_valid_time(college_end_time):
        college_end_time = DEFAULT_COLLEGE_END

    # --- Migration: break_after / break_periods ---
    break_after_raw = _get(raw, "break_after", "breakAfter")
    break_after = []
    if isinstance(break_after_raw, list):
        for p in break_after_raw:
            try:
                p_int = int(p)
            except (TypeError, ValueError):
                continue
            if 1 <= p_int < number_of_periods and p_int not in break_after:
                break_after.append(p_int)
        break_after.sort()
    else:
        # Try old model: break_periods
        break_periods_raw = _get(raw, "break_periods", "breakPeriods", default=[])
        if isinstance(break_periods_raw, list):
            for p in break_periods_raw:
                try:
                    p_int = int(p)
                except (TypeError, ValueError):
                    continue
                # Old model: break_period 3 meant "period 3 IS a break"
                # New model: break_after 2 means "break AFTER period 2"
                converted = p_int - 1
                if 1 <= converted < number_of_periods and converted not in break_after:
                    break_after.append(converted)
            break_after.sort()

    # --- Migration: lunch_after / lunch_break_period ---
    lunch_after_raw = _get(raw, "lunch_after", "lunchAfter")
    lunch_after = None
    if lunch_after_raw is not None:
        try:
            lunch_candidate = int(lunch_after_raw)
        except (TypeError, ValueError):
            lunch_candidate = None
        if lunch_candidate is not None and 1 <= lunch_candidate < number_of_periods:
            lunch_after = lunch_candidate
    else:
        # Try old model: lunch_break_period
        lunch_raw = _get(raw, "lunch_break_period", "lunchBreakPeriod")
        if lunch_raw is not None:
            try:
                lunch_candidate = int(lunch_raw) - 1
            except (TypeError, ValueError):
                lunch_candidate = None
            if lunch_candidate is not None and 1 <= lunch_candidate < number_of_periods:
                lunch_after = lunch_candidate

    stored_periods = _get(raw, "periods", default=None)
    periods = None
    if isinstance(stored_periods, list) and len(stored_periods) > 0:
        built = []
        for index, p in enumerate(stored_periods):
            if not isinstance(p, dict):
                built = None
                break
            start = p.get("start_time") or p.get("startTime")
            end = p.get("end_time") or p.get("endTime")
            if not _is_valid_time(start) or not _is_valid_time(end):
                built = None
                break
            built.append(
                {
                    "period_number": p.get("period_number") or p.get("periodNumber") or (index + 1),
                    "start_time": start,
                    "end_time": end,
                    "total_minutes": max(_time_to_minutes(end) - _time_to_minutes(start), 0),
                    "is_teaching": p.get("is_teaching", p.get("isTeaching", True)),
                    "is_break": p.get("is_break", p.get("isBreak", False)),
                    "is_lunch": p.get("is_lunch", p.get("isLunch", False)),
                }
            )
        periods = built

    if not periods:
        periods = derive_periods(
            college_start_time,
            period_duration,
            number_of_periods,
            break_after=break_after,
            lunch_after=lunch_after,
            break_time=break_time,
            lunch_time=lunch_time,
        )

    return {
        "working_days": working_days,
        "number_of_periods": number_of_periods,
        "college_start_time": college_start_time,
        "college_end_time": college_end_time,
        "period_duration": period_duration,
        "break_time": break_time,
        "lunch_time": lunch_time,
        "break_after": break_after,
        "lunch_after": lunch_after,
        "periods": periods,
    }


def instructional_period_numbers(config: Dict[str, Any]) -> List[int]:
    """
    Period numbers that are actual teaching slots. With the new model,
    we filter the periods list for is_teaching entries.
    """
    return [
        p["period_number"]
        for p in config.get("periods", [])
        if p.get("is_teaching", True)
    ]


async def load_config(db: AsyncSession, user_id: Optional[str]) -> Dict[str, Any]:
    """
    Load the user's Setup configuration from the database, falling back to
    defaults when absent. The user_id may be None (system/legacy callers); in
    that case defaults are returned.
    """
    if not user_id:
        return build_default_config()

    result = await db.execute(select(SetupConfig).where(SetupConfig.user_id == user_id))
    row = result.scalars().first()
    if row is None:
        return build_default_config()
    return normalize_config(row.config)


def validate_config_for_generation(config: Dict[str, Any]) -> None:
    """
    Raise a clear scheduling error when the Setup configuration cannot drive
    generation (missing working days or no usable teaching periods).
    """
    if not config.get("working_days"):
        raise SchedulingFailureError(
            "Please complete Setup configuration before generating the timetable."
        )
    if not config.get("periods") or not instructional_period_numbers(config):
        raise SchedulingFailureError(
            "Please complete Setup configuration before generating the timetable."
        )
